purge: add initial playbook

guits opened the pull request:
This playbook is intended for purging a whole Ceph cluster

Closes: https://bugzilla.redhat.com/show_bug.cgi?id=1881192

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
tests: add deploy-cluster.yml playbook

For CI purposes only.
It deploys a Ceph cluster with cephadm in order to test the purge
playbook.

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
purge: add initial playbook

This playbook is intended for purging a whole Ceph cluster

Closes: https://bugzilla.redhat.com/show_bug.cgi?id=1881192

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

sebastian-philipp commented on cephadm-purge-cluster.yml:
```diff
@@ -0,0 +1,59 @@
+
+    - name: set_fact container_binary
+      set_fact:
+        container_binary: "{{ 'podman' if (podman_binary.stat.exists and ansible_facts['distribution'] == 'Fedora') or (ansible_facts['os_family'] == 'RedHat' and ansible_facts['distribution_major_version'] == '8') else 'docker' }}"
```
can we avoid caring about this? You don't need to call `cephadm --docker rm-cluster`

sebastian-philipp commented on cephadm-purge-cluster.yml:
```diff
@@ -0,0 +1,59 @@
+        cephadm_cmd: "cephadm {{ '--docker' if container_binary == 'docker' else '' }}"
+
+    - name: get cluster fsid
+      command: "{{ cephadm_cmd }} shell -- ceph --cluster {{ cluster }} fsid"
```
Why is `--cluster {{ cluster }}` required here?

guits commented:
note to myself: still missing some tests after purge playbook is done

sebastian-philipp commented on library/cephadm_bootstrap.py:
```diff
@@ -0,0 +1,237 @@
+        description:
+            - Use docker instead of podman.
+        required: false
+    fsid:
+        description:
+            - Ceph FSID.
+        required: false
```
I'd like to avoid exposing the fsid here. I do not yet see any gain in letting users mess with the fsid. Can we avoid this?

guits replied:
this module was taken from ceph-ansible repo, it's intended to be used for CI purposes only here

sebastian-philipp replied:
hm. Makes sense for the CI. WDYT about removing this from the documentation

guits replied:
I was even thinking about not using it at all in fact.
I'm not sure it is worth shipping all of this only for ci

sebastian-philipp commented on library/cephadm_bootstrap.py:
```diff
@@ -0,0 +1,237 @@
+            mon_ip=dict(type='str', required=True),
+            image=dict(type='str', required=False),
+            docker=dict(type='bool', required=False, default=False),
+            fsid=dict(type='str', required=False),
```
remove?

guits replied:
ditto

sebastian-philipp commented on tests/functional/deploy-cluster.yml:
```diff
@@ -0,0 +1,280 @@
+        path: /usr/bin/podman
+      register: podman_binary
+
+    - name: set_fact container_binary
+      set_fact:
+        container_binary: "{{ 'podman' if (podman_binary.stat.exists and ansible_facts['distribution'] == 'Fedora') or (ansible_facts['os_family'] == 'RedHat' and ansible_facts['distribution_major_version'] == '8') else 'docker' }}"
```
```suggestion
    - name: set_fact container_binary
      set_fact:
        container_binary: "{{ 'podman' if podman_binary.stat.exists else 'docker' }}"
```
?

guits replied:
in fact, are we going to use docker instead of podman in the ci? (this playbook is here for CI testing only)

guits commented on cephadm-purge-cluster.yml:
```diff
@@ -0,0 +1,59 @@
+      command: "{{ cephadm_cmd }} shell -- ceph --cluster {{ cluster }} fsid"
+      register: cluster_fsid
+      run_once: true
+      when: inventory_hostname in groups.get(mon_group_name, []) or '_admin' in labels | default([])
```
this probably needs to be reworked

guits committed:
tests: add [admin] group in inventory

as per cephadm-purge.yml requirement, let's define an admin group.

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
tests: wait all osd are up before purging

Let's wait that services are up before purging the cluster.
Otherwise, for some reason it can make cephadm taking a very long time
to purge the cluster.

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits commented on cephadm-purge-cluster.yml:
```diff
@@ -0,0 +1,73 @@
+  become: true
+  gather_facts: false
+  tasks:
+    - import_role:
+        name: ceph-defaults
```
```suggestion
```

not needed

guits commented on cephadm-purge-cluster.yml:
```diff
@@ -0,0 +1,73 @@
+  become: true
+  gather_facts: false
+  tasks:
+    - import_role:
+        name: ceph-defaults
```
```suggestion
```

not needed

guits commented on cephadm-purge-cluster.yml:
```diff
@@ -0,0 +1,73 @@
+  hosts: localhost
+  become: false
+  tasks:
+    - import_role:
+        name: ceph-defaults
```
```suggestion
```

not needed

guits commented on cephadm-purge-cluster.yml:
```diff
@@ -0,0 +1,73 @@
+  become: true
+  gather_facts: false
+  tasks:
+    - import_role:
+        name: ceph-defaults
```
ditto

guits commented on cephadm-purge-cluster.yml:
```diff
@@ -0,0 +1,73 @@
+          You must have /etc/ceph/ceph.client.admin.keyring on this node ({{ inventory_hostname }})
+      when: not admin_keyring_stat.stat.exists | bool
+
+- name: pause cephadm
+  hosts: "admin[0]"
+  become: true
+  gather_facts: false
+  tasks:
```
```suggestion
```
no need to make a new play
